### PR TITLE
ovs_stats: Implement OVS statistics plugin.

### PR DESCRIPTION
**Pull request by maryamtahhan**
OVS stats plugin uses OVS utils to get available
bridge/interface statistics from OVSDB.

Change-Id: Iabe71d64fef7fdbfd5a272c3599a92bce2adf055
Signed-off-by: Mytnyk, VolodymyrX <volodymyrx.mytnyk@intel.com>

### CODE REVIEW

File: src/collectd.conf.pod [hunk 1]
```diff
@@ -5524,6 +5524,54 @@ use B<Interval> option of the OVS B<LoadPlugin> block settings. For milliseconds
 simple divide the time by 1000 for example if the desired interval is 50ms, set
 interval to 0.05.
 
+=head2 Plugin C<ovs_stats>
+
+The I<ovs_stats> plugin collects a statistics of OVS connected interfaces.
```
Comment (rubenk) [inline src/collectd.conf.pod]: statistics is plurar, so please remove the 'a'

File: src/ovs_stats.c [hunk 2]
```diff
@@ -0,0 +1,982 @@
...
+        [rx_broadcast_packets] = "rx_broadcast_packets",
+        [tx_broadcast_packets] = "tx_broadcast_packets",
+        [rx_undersized_errors] = "rx_undersized_errors",
+        [rx_oversize_errors] = "rx_oversize_errors",
+        [rx_fragmented_errors] = "rx_fragmented_errors",
+        [rx_jabber_errors] = "rx_jabber_errors",
+};
+
+/* Entry into the list of network bridges */
+static bridge_list_t *g_bridge_list_head = NULL;
```
Comment (rubenk) [inline src/ovs_stats.c]: no need to zero-initialize statics.

File: src/ovs_stats.c [hunk 3]
```diff
@@ -0,0 +1,982 @@
...
+};
+typedef struct ovs_stats_config_s ovs_stats_config_t;
+
+static ovs_stats_config_t ovs_stats_cfg = {
+    .ovs_db_node = "localhost", /* use default OVS DB node */
+    .ovs_db_serv = "6640",      /* use default OVS DB service */
+};
+
+static const iface_counter ovs_stats_counter_name_to_type(const char *counter) {
+  iface_counter index = not_supportred;
```
Comment (rubenk) [inline src/ovs_stats.c]: not_supported

File: src/ovs_stats.c [hunk 4]
```diff
@@ -0,0 +1,982 @@
...
+
+#include "common.h"
+
+#include "utils_ovs.h" /* OvS helpers */
+
+/* Plugin name */
+static const char plugin_name[] = "ovs_stats";
+
+typedef enum iface_counter {
+  not_supportred = -1,
```
Comment (rubenk) [inline src/ovs_stats.c]: not_supported

File: src/ovs_stats.c [hunk 5]
```diff
@@ -0,0 +1,982 @@
...
+                           OVS_DB_TABLE_CB_FLAG_INITIAL |
+                           OVS_DB_TABLE_CB_FLAG_INSERT |
+          OVS_DB_TABLE_CB_FLAG_MODIFY);
+}
+
+/* Check if bridge is configured to be monitored in config file */
+static int ovs_stats_is_monitored_bridge(const char *br_name) {
+  int rc = 0;
+  /* if no bridges are configured, return true */
+  if (!(rc = (g_monitored_bridge_list_head == NULL)))
```
Comment (rubenk) [inline src/ovs_stats.c]: It's a bit hard to follow this code. I believe if you can get rid of rc which you are treating as a bool this can be simplified a bit.

File: src/ovs_stats.c [hunk 6]
```diff
@@ -0,0 +1,982 @@
...
+    rc = (ovs_stats_get_bridge(g_monitored_bridge_list_head, br_name) != NULL);
+  return rc;
+}
+
+/* Delete all ports from port list */
+static void ovs_stats_free_port_list(port_list_t *head) {
+  port_list_t *i, *del;
```
Comment (rubenk) [inline src/ovs_stats.c]: can you move these into the for loop?

File: src/ovs_stats.c [hunk 7]
```diff
@@ -0,0 +1,982 @@
...
+  for (i = head; i != NULL;) {
+    del = i;
+    i = i->next;
+    sfree(del);
+  }
+}
+
+/* Delete all bridges from bridge list */
+static void ovs_stats_free_bridge_list(bridge_list_t *head) {
+  bridge_list_t *i, *del;
```
Comment (rubenk) [inline src/ovs_stats.c]: same here

File: src/ovs_stats.c [hunk 8]
```diff
@@ -0,0 +1,982 @@
...
+    del = i;
+    i = i->next;
+    sfree(del->name);
+    sfree(del);
+  }
+}
+
+/* Handle OVSDB lost connection callback */
+static void ovs_stats_conn_terminate() {
+
```
Comment (rubenk) [inline src/ovs_stats.c]: remove blank linke

File: src/ovs_stats.c [hunk 9]
```diff
@@ -0,0 +1,982 @@
...
+  for (i = head; i != NULL;) {
+    del = i;
+    i = i->next;
+    sfree(del->name);
+    sfree(del);
+  }
+}
+
+/* Handle OVSDB lost connection callback */
+static void ovs_stats_conn_terminate() {
```
Comment (rubenk) [inline src/ovs_stats.c]: (void)

File: src/ovs_stats.c [hunk 10]
```diff
@@ -0,0 +1,982 @@
...
+/* Initialize OvS Stats plugin*/
+static int ovs_stats_plugin_init(void) {
+  ovs_db_callback_t cb = {.post_conn_init = ovs_stats_initialize,
+                          .post_conn_terminate = ovs_stats_conn_terminate};
+
+  INFO("%s: Connecting to OVS DB using address=%s, service=%s, unix=%s",
+       plugin_name, ovs_stats_cfg.ovs_db_node, ovs_stats_cfg.ovs_db_serv,
+       ovs_stats_cfg.ovs_db_unix);
+  /* connect to OvS DB */
+  if ((ovs_db = ovs_db_init (ovs_stats_cfg.ovs_db_node,
```
Comment (rubenk) [inline src/ovs_stats.c]: ovs_db is static. You're using the g_ prefix for all other statics, I believe you should do the same here.

**Comment by rubenk** [comment]
@maryamtahhan ping?

**Comment by taraschornyi** [comment]
@rubenk sorry for long reply will update this PR today or tomorrow

**Comment by rubenk** [comment]
Thank you @maryamtahhan and @taraschornyi.
